odoo/tools/translate.py

```diff
@@ -1197,177 +1197,189 @@ def _export_translatable_resources(self):
-    :param lang: language code of the translations contained in `fileobj`
-                 language must be present and activated in the database
-    :param verbose: increase log output
-    :param overwrite: if a translation already exists for a term, replace it with
-                      the one in `fileobj`
-    """
-    if verbose:
-        _logger.info('loading translation file for language %s', lang)
-    env = odoo.api.Environment(cr, odoo.SUPERUSER_ID, {'lang': lang})
-    if not env['res.lang']._lang_get(lang):
-        _logger.error("Couldn't read translation for lang '%s', language not found", lang)
-        return None
-    try:
-        fileobj.seek(0)
-        reader = TranslationFileReader(fileobj, fileformat=fileformat)
-        _trans_load_data(cr, reader, lang, overwrite)
-
-        if verbose:
-            _logger.info("translation file loaded successfully")
-
-    except IOError:
-        iso_lang = get_iso_codes(lang)
-        filename = '[lang: %s][format: %s]' % (iso_lang or 'new', fileformat)
-        _logger.exception("couldn't read translation file %s", filename)
-
-
-def _trans_load_data(cr, reader, lang, overwrite=False, force_overwrite=False, xml_ids=None):
-    # logical overwrite: if a translation should be updated when it has already been in the DB before
-    # logical overwrite = force_overwrite or (not noudpate and overwrite)
-    env = odoo.api.Environment(cr, odoo.SUPERUSER_ID, {'lang': lang})
-    if force_overwrite:
-        overwrite = True
-
-    # load all translation in to dict
-    deep_defaultdict = lambda: defaultdict(deep_defaultdict)
-    # {model_name: {xmlid: {field_name, {src: value}}}}
-    translations = deep_defaultdict()
-    # [(module_name, imd_name), (module_name, imd_name), ...] only for model_terms translation
-    module_imd_names = []
-    # {model_name: {field_name: [xmlid, xmlid, ...]}}
-    model_name_to_field_name_to_xmlids = defaultdict(lambda: defaultdict(list))
-
-    for row in reader:
-        if not row.get('value') or not row.get('src'):  # ignore empty translations
-            continue
-        if row.get('type') == 'code':  # ignore code translations
-            continue
-        # TODO: CWG if the po file should not be trusted, we need to check each model term
-        model_name = row.get('imd_model')
-        module_name = row['module']
-        imd_name = row['imd_name']
-        if model_name in env:
+class TranslationImporter:
+    def __init__(self, cr, verbose=True):
+        self.cr = cr
+        self.verbose = verbose
+        self.env = odoo.api.Environment(cr, odoo.SUPERUSER_ID, {})
+
+        # {model_name: {field_name: {xmlid: {lang: value}}}}
+        self.model_translations = deep_defaultdict()
+        # {model_name: {field_name: {xmlid: {src: {lang: value}}}}}
+        self.model_terms_translations = deep_defaultdict()
+
+    def load(self, fileobj, fileformat, lang, xmlids=None):
```

Comment:
`fileobj` -> `file` agree

for `load` and `load_file`
I find usually, python projects like to use `function_name_file` for `filepath`
and `function_name` without `_file` usually accepts fileobj
However, some python projects (e.g. `pip`) use `def read/write(filepath)` and `def read_file/write_file(fileob)`

I would say it is better to avoid using `_file` to distinguish between the two
And I get the idea from @xmo-odoo , maybe we could only support api for file object

Reply:
We _can_ support the file object API only, but supporting the filename alternative is actually valuable (in terms of caller code simplicity and readability).

Reply:
OK, agree, supporting both apis is convenient.
However I suggest to use
```
translation_importer.load_file(filename, lang, xmlids)
translation_importer.load(file, fileformat, lang, xmlids)
```
I think it is closer to what we have in the code (e.g. load_information_from_description_file, load_test_file_py, handle_file) which use `_file` to denote the filename/filepath str in the parameters